### DISCUSSION TITLE
Support for APNGs?

### DISCUSSION THREAD

**Discussion by JGVB** [libvips/libvips#2000] [Ideas]
APNG's are on the rise and I am wondering if there is support for them in this library?  Or if there could be with any modifications?

**Comment by jcupitt** [comment]
Sorry, no, it would need to be added. I think this is usually done as a modification to the png reader, which is here:

https://github.com/jcupitt/libvips/blob/master/libvips/foreign/vipspng.c

I'll tag this issue as "enhancement".

**Comment by JGVB** [comment]
I am assuming this would be non-trivial to accomplish?

**Comment by jcupitt** [comment]
Yes, it would take some work. 

From a quick read it seems that APNG is not an official PNG format, so the standard `libpng` library does not support it. Instead, you need to build your own `libpng` with a set of patches, then also add code to your PNG writer to drive the extra parts of the library. 

Maybe when/if there's a `libapng` for this format it'll become easier to support.

**Comment by adamdilek** [comment]
It will be good.

**Comment by JGVB** [comment]
Actually figured out a way using APNGasm. (https://sourceforge.net/projects/apngasm/) Basically, you disassemble the apng into it's various frames, resize them all, then reassemble them

**Comment by adamdilek** [comment]
Thanks @JGVB for answer. We are using same way at @mojilala. The problem is `delay` and `loop` options. Do you know any way to find apng file's loop ?

**Comment by JGVB** [comment]
@muhammet  I'm sorry, I'm not sure what you mean by the apng file's loop.  Are you referring to the options you pass into the software?  I'm pretty sure the default arguments are fine.

**Comment by adamdilek** [comment]
Yes I'm referring to the options you pass into the software.

On Thu, Jan 12, 2017 at 10:57 PM, James GVB <notifications@github.com>
wrote:

> @muhammet <https://github.com/muhammet> I'm sorry, I'm not sure what you
> mean by the apng file's loop. Are you referring to the options you pass
> into the software? I'm pretty sure the default arguments are fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jcupitt/libvips/issues/562#issuecomment-272281273>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA8aP_3tYtCpnBFseB91dYRYiQsESfkOks5rRpO4gaJpZM4K4k9P>
> .
>



-- 
Muhammet

**Comment by LudwigvonWu** [comment]
I recommend looking at the specification here:

https://wiki.mozilla.org/APNG_Specification

The loop count is found in the Animation Control Chunk (acTL), for the delay of each frame, it's contained in the Frame Control Chunk (fcTL), but the APNG Disassembler automatically reads it and writes a txt file with said information, having the same name of the extracted frame.

**Comment by aopal** [comment]
Resurrecting this thread as APNG has become much more popular since the initial discussion; most major browsers now support it, as do other tools like ImageMagick and ffmpeg. 

I'd love to see APNG support in libvips. Perhaps when libspng adds support? https://github.com/randy408/libspng/projects/3

**Reply by jcupitt** [reply]
OK, I made an enhancement issue for it.

It was tricky to support before since you had to add extra stuff on top of libpng, but once libspng has support baked in, it should be relatively easy.

**Comment by jcupitt** [comment]
Enhancement issue, might be in 8.16 

https://github.com/libvips/libvips/issues/2537
